src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts

```diff
@@ -0,0 +1,87 @@
+import * as btc from '@scure/btc-signer';
+
+import { isDefined } from '@shared/utils';
+
+import { PsbtInputForUi } from './use-psbt-decoded-request';
+
+export interface InputOutputPair {
+  inputs: PsbtInputForUi[];
+  output: btc.TransactionOutputRequired;
+}
+
+function consolidateInputsWithSameAddress(inputs: PsbtInputForUi[]): PsbtInputForUi[] {
+  const utxo: Record<string, PsbtInputForUi> = {};
+  inputs.forEach(({ address, value, ...rest }) => {
+    if (isDefined(utxo[address])) {
+      utxo[address] = {
+        ...utxo[address],
+        value: utxo[address].value + value,
+      };
+    } else {
+      utxo[address] = { address, value, ...rest };
+    }
+  });
+  return Object.values(utxo);
+}
+
+interface MatchInputsOutputsArgs {
+  psbtInputs: PsbtInputForUi[];
+  unsignedOutputs: btc.TransactionOutputRequired[];
+}
+export function matchInputsOutputs({ psbtInputs, unsignedOutputs }: MatchInputsOutputsArgs): {
+  fee: number;
+  inputOutputPairs: InputOutputPair[];
+} {
+  const pairs: InputOutputPair[] = [];
+  const availableInputs = [...psbtInputs];
+  let remainingInputValue = 0;
+
+  for (const output of unsignedOutputs) {
+    const inputOutputPair: InputOutputPair = {
+      inputs: [],
+      output: output,
+    };
+
+    // Calculate the remaining value needed for this output
+    let remainingOutputValue = Number(output.amount);
+
+    // Iterate through the available inputs and add them until the output is fulfilled
+    for (let i = 0; i < availableInputs.length; i++) {
+      const input = availableInputs[i];
+
+      if (remainingInputValue === 0) remainingInputValue = input.value;
+
+      // Handle if the unsigned input at index has an inscription
+      if (availableInputs[i].unsignedUtxo?.inscriptions) {
+        inputOutputPair.inputs.push(input);
+        remainingInputValue = 0;
+        availableInputs.splice(i, 1);
+        break;
```

Comment:
It just breaks it for that output bc it has been fulfilled, but continues with other pairing. I have only seen one input to an output for an inscription, is that not always the case?

Reply:
Maybe a comment here would help explain what's going on, I'm not really sure what it does. 

Input/outputs can be many-to-one, one-to-many, many-to-many, and referring to them as "pairs" doesn't align with that.

I think the complexity of this feature lies in the blur between building a "psbt signing feature" vs "inscription marketplace transfers (that just happens to use psbts)". Ideally, inscription features are somehow layered on top of an inscription-agnostic psbt-imlpementation.

Reply:
How would this currently handle the many-to-many scenario in which you have 1 input from address X that produces outputs to addresses A,B,C, etc.?

Presumably we'd see a 10 groupings of "Output/Input" with the input data shown in a redundant way? Or would something else happen?

If the former, it may not be the most ideal UX but seems viable for an iterative enhancement here.

Reply:
I just removed handling the inscription in the matching and seems to be ok, right? Now multiple inputs cover the output...

![Screen Shot 2023-04-12 at 12 16 47 PM](https://user-images.githubusercontent.com/6493321/231534305-e3d24244-235a-41ab-821e-9baf64c106d8.png)

Reply:
This was always working when an input wasn't an inscription, but I had overlooked would be the same with an inscription.